funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py

```diff
@@ -0,0 +1,108 @@
+        self._deliveries = None
+        self._channel = None
+        self._stopping = False
+
+    def connect(self) -> pika.BlockingConnection:
+        """Connect
+
+        :rtype: pika.BlockingConnection
+        """
+        logger.info(f"Connecting to {self.params}")
+        self._conn = pika.BlockingConnection(self.params)
+        self._channel = self._conn.channel()
+        self._channel.confirm_delivery()
+        self._channel.exchange_declare(
+            exchange=self.exchange, exchange_type=self.exchange_type
+        )
+        # TO-DO: This shouldn't be done on the endpoint side
+        self._channel.queue_declare(queue="results")
+        self._channel.queue_bind(
+            queue="results", exchange=self.exchange, routing_key="*.results"
+        )
+        self._is_connected = True
+
+        return self._conn
+
+    def publish_message_retryable(self, message: bytes, retry_count=0, max_retries=3):
+        """Publish message to RabbitMQ with the routing key to identify the message source
+        The channel specifies confirm_delivery and with `mandatory=True` this call
+        will *block* until a delivery confirmation is received.
+
+        """
+        try:
+            self._channel.basic_publish(
+                self.exchange, self.routing_key, message, mandatory=True
+            )
+        except pika.exceptions.UnroutableError:
+            logger.exception("Message could not be delivered.")
+            raise
+        except (
+            pika.exceptions.ConnectionClosedByBroker,
+            pika.exceptions.ConnectionClosed,
+            pika.exceptions.ConnectionBlockedTimeout,
+        ):
+            logger.exception("Disconnected from the broker, attempting reconnect")
+            if retry_count < max_retries:
+                time.sleep(3 ** retry_count)  # Hacky exponential back-off
+                try:
+                    self.connect()
+                except Exception:
+                    logger.warning(
+                        f"Failed to reconnect, attempt {retry_count}/{max_retries}"
+                    )
+                    return self.publish(message, retry_count=retry_count + 1)
+            else:
+                raise
+
+    def publish(self, message: bytes) -> None:
+        """Publish message to RabbitMQ with the routing key to identify the message source
+        The channel specifies confirm_delivery and with `mandatory=True` this call
+        will *block* until a delivery confirmation is received.
+
+        Raises: Exception from pika if the message could not be delivered
+
+        """
+        try:
+            self._channel.basic_publish(
+                self.exchange, self.routing_key, message, mandatory=True
+            )
+        except Exception:
+            logger.exception("Message could not be delivered")
+            raise
+
+    def close(self):
+        self._channel.close()
+        self._conn.close()
+        self._is_connected = False
```

Comment:
I'm not familiar with the pika/rabbitmq semantics for a channel. Do we need to close the channel when we are done using it?

If this is required, it might be better to make this behave as a context manager by defining `__enter__` and `__exit__`.